Objects/regions.c

```diff
@@ -88,6 +89,168 @@ bool is_c_wrapper(PyObject* obj){
+    return Py_None;
+}
+
+/**
+ * Set the global variables for a failure.
+ * This allows the interpreter to inspect what has failed.
+ */
+void set_failed_edge(PyObject* src, PyObject* tgt)
+{
+    Py_IncRef(src);
+    error_src = src;
+    Py_IncRef(tgt);
+    error_tgt = tgt;
+    printf("Error: %p -> %p: destination is not immutable\n", src, tgt);
+    // We have discovered a failure.
+    // Disable region check, until the program switches it back on.
+    do_region_check = false;
+    error_occurred = true;
+}
+
+PyObject* _Py_InvariantSrcFailure(void)
+{
+    return Py_NewRef(error_src);
+}
+
+PyObject* _Py_InvariantTgtFailure(void)
+{
+    return Py_NewRef(error_tgt);
+}
+
+
+// Lifted from gcmodule.c
+typedef struct _gc_runtime_state GCState;
+#define GEN_HEAD(gcstate, n) (&(gcstate)->generations[n].head)
+#define GC_NEXT _PyGCHead_NEXT
+#define GC_PREV _PyGCHead_PREV
+#define FROM_GC(g) ((PyObject *)(((char *)(g))+sizeof(PyGC_Head)))
+
+
+/* A traversal callback for _Py_CheckRegionInvariant.
+   - op is the target of the reference we are checking, and
+   - parent is the source of the reference we are checking.
+*/
+static int
+visit_invariant_check(PyObject *op, void *parent)
+{
+    PyObject *src_op = _PyObject_CAST(parent);
+    // Check Immutable only reaches immutable
+    if ((src_op->ob_region == _Py_IMMUTABLE)
+        && (op->ob_region != _Py_IMMUTABLE))
+        {
+            set_failed_edge(src_op, op);
+            return 0;
+        }
+    // TODO: More checks to go here as we add more region
+    // properties.
+
+    return 0;
+}
+
+/**
+ * This uses checks that the region topology is valid.
+ *
+ * It is currently implemented using the GC data. This
+ * means that not all objects are traversed as some objects
+ * are considered to not participate in cycles, and hence
+ * do not need to be understood for the cycle detector.
+ *
+ * This is not ideal for the region invariant, but is a good
+ * first approximation.  We could actually walk the heap
+ * in a subsequent more elaborate invariant check.
+ *
+ * Returns non-zero if the invariant is violated.
+ */
+int _Py_CheckRegionInvariant(PyThreadState *tstate)
+{
+    // Check if we should perform the region invariant check
+    if(!do_region_check){
+        return 0;
+    }
+
+    // Use the GC data to find all the objects, and traverse them to
+    // confirm all their references satisfy the region invariant.
+    GCState *gcstate = &tstate->interp->gc;
+
+    // There is an cyclic doubly linked list per generation of all the objects
+    // in that generation.
+    for (int i = NUM_GENERATIONS-1; i >= 0; i--) {
+        PyGC_Head *containers = GEN_HEAD(gcstate, i);
+        PyGC_Head *gc = GC_NEXT(containers);
+        // Walk doubly linked list of objects.
+        for (; gc != containers; gc = GC_NEXT(gc)) {
+            PyObject *op = FROM_GC(gc);
+            // Local can point to anything.  No invariant needed
+            if (op->ob_region == _Py_DEFAULT_REGION)
```

Comment:
Question: Is the current local region just a constant? Meaning it not bound to the current thread or sub interpreter?

Reply:
Currently a constant.  This will need changing, but I think the changes should be fairly easy.